### DISCUSSION TITLE
Fixing Jumper addon

### DISCUSSION THREAD

**Discussion by euu2021** [freeplane/freeplane#374] [Help] [answered]
Hi,

@lilive created a brilliant addon called [Jumper](https://github.com/lilive/Freeplane-Jumper), but it has a problem that makes it unusable with undocked windows: the Jumper window opens at the main window and not on the undocked window. 

His other addon called Bookmarks also had this problem, but he fixed it. The fix, as he said himself, [was simple](https://github.com/lilive/Freeplane-Bookmarks-add-on/commit/1a132ea97458c67a72860b9aefab886b8ce5045c): replacing ui.frame for ui.currentFrame in the code.

I'm trying to apply the same fix to the Jumper addon. I made the replacements where I found ui.frame in the groovy files, and now I need to rebuild the addon. I'm following [the steps given by lilive](https://github.com/lilive/Freeplane-Jumper#compilation-from-the-sources), but the jar building is failing with this exception:

![image](https://user-images.githubusercontent.com/77707706/168798405-17e13348-4800-4fd9-b412-7bf294340e6c.png)

How should I proceed?

Please, help.
Thanks,

EDIT: I forgot to mention that lilive is not updating his addons anymore, so I cannot ask him to fix it.

**Comment by Illegit** [comment]
Hi eu2021, even I'm a user of the jumper and very much appreciate your effort to correct it. Thank you.
illegit

**Comment by EdoFro** [comment, marked as answer]
Hi @eu2021,
I think you have to edit the build.gradle file and edit the folder path defined for Freeplane so it matches the folder path where Freeplane is installed in your computer.

this lines [here](https://github.com/lilive/Freeplane-Jumper/blob/0098f7d3b8f40a06d88b807b5295d5ea865e6c66/build.gradle#L12-L14):
```

        "C:/Program Files/Freeplane",
        "C:/Program Files/Freeplane/core/org.freeplane.core/lib",
		"C:/Program Files/Freeplane/plugins/org.freeplane.plugin.script/lib"

```

hope it helps,

edo

**Reply by EdoFro** [reply]
Or it may be your Gradle version. the `compile` command was deprecated. Now you have to use `implementation`. Just do a Find & Replace and it should work.

Other point to consider:  Freeplane v1.10 comes with Groovy 4, maybe there is a change you have to consider too.

**Reply by EdoFro** [reply]
in that case you have to change this part:
```groovy
dependencies {
    compile name: 'freeplanelauncher'
    compile name: 'freeplaneviewer'
	compile group: 'org.freeplane.script', name: 'plugin', version: '+'
	compile ('org.codehaus.groovy:groovy-all:3.0.1') {
		exclude group:'org.codehaus.groovy', module:'groovy-test'
		exclude group:'org.codehaus.groovy', module:'groovy-test-junit5'
		exclude group:'org.codehaus.groovy', module:'groovy-testng'
		exclude group:'org.codehaus.groovy', module:'groovy-ant'
		exclude group:'org.codehaus.groovy', module:'groovy-docgenerator'
		exclude group:'org.codehaus.groovy', module:'groovy-groovydoc'
		exclude group:'org.codehaus.groovy', module:'groovy-cli-commons'
		exclude group:'org.codehaus.groovy', module:'groovy-cli-picocli'
	}
	compile 'org.codehaus.groovy:groovy-dateutil:3.0.1'
}
```

changing:

- because of Gradle version:
   - compile --> implementation

- because of Groovy version:
   - org.codehaus.groovy --> org.apache.groovy
   - 3.0.1  -->  4.0.1

(not tested):
```groovy
dependencies {
    implementation name: 'freeplanelauncher'
    implementation name: 'freeplaneviewer'
	implementation group: 'org.freeplane.script', name: 'plugin', version: '+'
	implementation ('org.apache.groovy:groovy-all:4.0.1') {
		exclude group:'org.apache.groovy', module:'groovy-test'
		exclude group:'org.apache.groovy', module:'groovy-test-junit5'
		exclude group:'org.apache.groovy', module:'groovy-testng'
		exclude group:'org.apache.groovy', module:'groovy-ant'
		exclude group:'org.apache.groovy', module:'groovy-docgenerator'
		exclude group:'org.apache.groovy', module:'groovy-groovydoc'
		exclude group:'org.apache.groovy', module:'groovy-cli-commons'
		exclude group:'org.apache.groovy', module:'groovy-cli-picocli'
	}
	implementation 'org.apache.groovy:groovy-dateutil:4.0.1'
}
```

**Reply by euu2021** [reply]
the "implement"change was good. But, I'm having problems with the groovy. I will explain later.

**Reply by EdoFro** [reply]
Now I have some free time if you need some help

**Reply by euu2021** [reply]
I tried with Freeplane-Setup-1.9.14.exe, but Gradle was asking for some groovy jars. I found some of them on the internet and threw in the FP folder. But, there are still those that I can't find on the internet (the download page goes 404):

```
* What went wrong:
Execution failed for task ':compileGroovy'.
> Could not resolve all files for configuration ':compileClasspath'.
   > Could not find org.apache.groovy:groovy-all:3.0.1.
     Searched in the following locations:
       - https://repo.maven.apache.org/maven2/org/apache/groovy/groovy-all/3.0.1/groovy-all-3.0.1.pom
       - file:/C:/Program Files/Freeplane/groovy-all-3.0.1.jar
       - file:/C:/Program Files/Freeplane/groovy-all.jar
       - file:/C:/Program Files/Freeplane/core/org.freeplane.core/lib/groovy-all-3.0.1.jar
       - file:/C:/Program Files/Freeplane/core/org.freeplane.core/lib/groovy-all.jar
       - file:/C:/Program Files/Freeplane/plugins/org.freeplane.plugin.script/lib/groovy-all-3.0.1.jar
       - file:/C:/Program Files/Freeplane/plugins/org.freeplane.plugin.script/lib/groovy-all.jar
     Required by:
         project :
```

Now, I'm trying with Freeplane-Setup-with-Java-1.10.3-pre01.exe, and it's giving me this giant complaint:

```
* What went wrong:
Execution failed for task ':compileGroovy'.
> Could not resolve all files for configuration ':compileClasspath'.
   > Could not resolve org.apache.groovy:groovy-all:4.0.1.
     Required by:
         project :
      > No matching variant of org.apache.groovy:groovy-all:4.0.1 was found. The consumer was configured to find an API of a library compatible with Java 8, preferably not packaged as a jar, preferably optimized for standard JVMs, and its dependencies declared externally but:
          - Variant 'apiElements' capability org.apache.groovy:groovy-all:4.0.1 declares an API of a component:
              - Incompatible because this component declares a platform and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about how its dependencies are found (required its dependencies declared externally)
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
                  - Doesn't say anything about its target Java version (required compatibility with Java 8)
                  - Doesn't say anything about its elements (required them preferably not packaged as a jar)
          - Variant 'enforcedApiElements' capability org.apache.groovy:groovy-all-derived-enforced-platform:4.0.1 declares an API of a component:
              - Incompatible because this component declares an enforced platform and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about how its dependencies are found (required its dependencies declared externally)
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
                  - Doesn't say anything about its target Java version (required compatibility with Java 8)
                  - Doesn't say anything about its elements (required them preferably not packaged as a jar)
          - Variant 'enforcedRuntimeElements' capability org.apache.groovy:groovy-all-derived-enforced-platform:4.0.1 declares a runtime of a component:
              - Incompatible because this component declares an enforced platform and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about how its dependencies are found (required its dependencies declared externally)
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
                  - Doesn't say anything about its target Java version (required compatibility with Java 8)
                  - Doesn't say anything about its elements (required them preferably not packaged as a jar)
          - Variant 'groovydocElements' capability org.apache.groovy:groovy-all:4.0.1 declares a runtime of a component, and its dependencies declared externally:
              - Incompatible because this component declares documentation and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
                  - Doesn't say anything about its target Java version (required compatibility with Java 8)
                  - Doesn't say anything about its elements (required them preferably not packaged as a jar)
          - Variant 'javadocElements' capability org.apache.groovy:groovy-all:4.0.1 declares a runtime of a component, and its dependencies declared externally:
              - Incompatible because this component declares documentation and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
                  - Doesn't say anything about its target Java version (required compatibility with Java 8)
                  - Doesn't say anything about its elements (required them preferably not packaged as a jar)
          - Variant 'runtimeElements' capability org.apache.groovy:groovy-all:4.0.1 declares a runtime of a component:
              - Incompatible because this component declares a platform and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about how its dependencies are found (required its dependencies declared externally)
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
                  - Doesn't say anything about its target Java version (required compatibility with Java 8)
                  - Doesn't say anything about its elements (required them preferably not packaged as a jar)
          - Variant 'sourcesElements' capability org.apache.groovy:groovy-all:4.0.1 declares a runtime of a component, and its dependencies declared externally:
              - Incompatible because this component declares documentation and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
                  - Doesn't say anything about its target Java version (required compatibility with Java 8)
                  - Doesn't say anything about its elements (required them preferably not packaged as a jar)
```

**Reply by EdoFro** [reply]
you don't need to install Groovy. It's allready installed with Freeplane.

look in your Freeplane directory into `Freeplane/plugins/org.freeplane.plugin.script/lib`

this is why it is specified [in this line in the build.gradle file](https://github.com/lilive/Freeplane-Jumper/blob/0098f7d3b8f40a06d88b807b5295d5ea865e6c66/build.gradle#L14) 

I looked at my Freeplane installation in my workplace notebook and it's v1.9.15.
In the `Freeplane/plugins/org.freeplane.plugin.script/lib` directory are all the groovy jar files, but the Groovy version is 3.0.8 , look in your computer and maybe that is the number version you have to specify.

This is how that directory looks in my case:
![image](https://user-images.githubusercontent.com/35700575/168890504-39a9a8de-a843-40d4-a74a-7385ceab1d0d.png)

**Reply by EdoFro** [reply]
in that case, the build.gradle file should end like this:

```groovy
dependencies {
    implementation name: 'freeplanelauncher'
    implementation name: 'freeplaneviewer'
	implementation group: 'org.freeplane.script', name: 'plugin', version: '+'
	implementation ('org.codehaus.groovy:groovy-all:3.0.8') {
		exclude group:'org.codehaus.groovy', module:'groovy-test'
		exclude group:'org.codehaus.groovy', module:'groovy-test-junit5'
		exclude group:'org.codehaus.groovy', module:'groovy-testng'
		exclude group:'org.codehaus.groovy', module:'groovy-ant'
		exclude group:'org.codehaus.groovy', module:'groovy-docgenerator'
		exclude group:'org.codehaus.groovy', module:'groovy-groovydoc'
		exclude group:'org.codehaus.groovy', module:'groovy-cli-commons'
		exclude group:'org.codehaus.groovy', module:'groovy-cli-picocli'
	}
	implementation 'org.codehaus.groovy:groovy-dateutil:3.0.8'
}
```

explanation:

- because of Gradle version:
   - compile --> implementation

- because of Groovy version:
   - 3.0.1 --> 3.0.8
   - DON'T :  org.codehaus.groovy -X-> org.apache.groovy (this is only if Groovy version >= 4)

**Reply by EdoFro** [reply]
in my Github repository I have some build.gradle files that works. For example [this one](https://github.com/EdoFro/Freeplane_Menu-o-Matic/blob/main/Menu-o-Matic/build.gradle). It may help you as an example.

In essence it's the same file, but my version is a little more complicated because I put in the same file the parameters to build it for Freeplane v1.9 and Freeplane 1.10
I was testing with it because I had the same problems that you are now suffering.

It works this way:
in [these lines](https://github.com/EdoFro/Freeplane_Menu-o-Matic/blob/bfff715858e53d66b6108d273bf3bd46a7ec050d/Menu-o-Matic/build.gradle#L10-L20), I define the parameters to be used:

```
def FPversion = 'v10_1'
def freeplaneDirs = [
        instaled     : "C:/Program Files/Freeplane",
        protableApps : "C:/PortableApps/FreeplanePortable/App/Freeplane",
        v9_15         : "C:/Freeplane/freeplane-1.9.15-pre01",
        v10_1         : "C:/Freeplane/freeplane-1.10.2"
    ]
def freeplaneDir = freeplaneDirs[FPversion]


def groovyGroupId = FPversion.startsWith('v10')?'org.apache.groovy':'org.codehaus.groovy'
def groovyVersion = FPversion.startsWith('v10')?'4.0.1':'3.0.8'
```

variable `freeplaneDirs ` contains a map with all directories where I have different versions of Freeplane installed.

variable `FPversion` tells which Freeplane version I'm going to use for the building process. I change that value and save the fiel before using Gradle.

variable `freeplaneDir` gets the path to use.

`groovyGroupId` and `groovyVersion` get the groovy definition depending on the version on FP to be used.


Your file doesn't need to be this way, I explain all this so you can understand a file that works with FP 1.9 and/or FP 1.10
bye,
edo

**Reply by EdoFro** [reply]
my file has also an extra line that I don't really know if it is necessary. But just in case: [line 46](https://github.com/EdoFro/Freeplane_Menu-o-Matic/blob/bfff715858e53d66b6108d273bf3bd46a7ec050d/Menu-o-Matic/build.gradle#L46)

**Reply by euu2021** [reply]
Yes, that was the problem. Thanks

I cooked a jar!!

![image](https://user-images.githubusercontent.com/77707706/168901044-96c3b1fe-3703-41be-9d74-8b44b379a26f.png)

Now, I follow with the next steps.

**Comment by euu2021** [comment]
[jumper-v1.0.1.addon (unofficial fixed version).zip](https://github.com/freeplane/freeplane/files/8712163/jumper-v1.0.1.addon.unofficial.fixed.version.zip)

It works!! 🚀😮

I will send a Pull Request to lilive, but I think he won't see it.
